### PR TITLE
Add URL handling to chat

### LINKED ISSUE #806: URL parsing support
We need proper support for URL parsing.

We may be able to borrow some code for this from osu-stable, since it was quite recently rewritten and have unit tests.

### PR DESCRIPTION
**Pull request by FreezyLemon**
Closes #806.

- [x] Depends on https://github.com/ppy/osu-framework/pull/1273

The old `LinkFormatter` class was mostly copied over and renamed to `MessageFormatter`.
It's implemented to format messages on `ChatLine.Message`'s setter.

For the actual link sprite, I took inspiration in the `OsuLinkSpriteText` class that was in ProfileHeader. I extended it and let `ChatLink` inherit it, which does all the stuff regarding click events and tooltips in chat.
The click events are not complete since not everything is implemented yet. So at the moment, some links will give a notification saying exactly that (see `OsuGame`).
I added a new `GetBeatmapRequest` which gets a map by map ID (not set ID), so that I could get the set ID off of that.

I also added a new test case and fixed the old `ChatDisplay` test case since it would break without the necessary dependencies.
The new interface `ICanDisableHoverSounds` adds a bool which lets a Parent element decide on whether to play the hover sound or not. This is needed so that the hover sound is not played too often (every word is a sprite, so if a link is multiple words long it would play the hover sound every time you move between words). If the interface is not implemented however, it is just ignored and the sound is played regardless (this is done for ease of use and not to break existing implementations).

The change in `PreviewButton` is something I found in testing: Sometimes preview.Length can be 0 (i don't quite remember why this happens), and when it is, it breaks its own Width (NaN) and crashes.

### CODE REVIEW

File: osu.Game/Graphics/UserInterface/HoverSounds.cs [hunk 1]
```diff
@@ -50,4 +51,13 @@ public enum HoverSampleSet
         [Description("-softer")]
         Soft
     }
+
+    /// <summary>
+    /// Classes implementing this interface can choose whether or not the HoverSounds should be played.
+    /// <para>If this is not implemented, the sounds will always be played when OnHover is triggered.</para>
+    /// </summary>
+    public interface ICanDisableHoverSounds
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/HoverSounds.cs]: you should find a way to do this without doing this

File: osu.Game/Online/Chat/ChatLink.cs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+        protected override Container<Drawable> Content => content ?? base.Content;
+
+        protected override void OnLinkClicked()
+        {
+            var url = Url;
+
+            if (url.StartsWith("osu://"))
+            {
+                url = url.Substring(6);
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatLink.cs]: please no magic substrings

File: osu.Game/Online/Chat/ChatLink.cs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                else
+                    base.OnLinkClicked();
+            }
+            else
+                base.OnLinkClicked();
+        }
+
+        private int getIdFromUrl(string url)
+        {
+            var lastSlashIndex = url.LastIndexOf('/');
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatLink.cs]: this should probably be regex

File: osu.Game/Graphics/Sprites/OsuLinkSpriteText.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+namespace osu.Game.Graphics.Sprites
+{
+    public class OsuLinkSpriteText : OsuSpriteText
+    {
+        protected override IEnumerable<Drawable> FlowingChildren => Children;
+
+        protected override bool OnClick(InputState state)
+        {
+            OnLinkClicked();
+            return base.OnClick(state);
```
Comment (peppy) [inline osu.Game/Graphics/Sprites/OsuLinkSpriteText.cs]: shouldn't this return true?

File: osu.Game/Overlays/BeatmapSet/PreviewButton.cs [hunk 5]
```diff
@@ -80,7 +80,7 @@ protected override void Update()
         {
             base.Update();
 
-            if (Playing.Value && preview != null)
+            if (Playing.Value && preview != null && preview.Length > 0)
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapSet/PreviewButton.cs]: what is this doing in this PR?

Comment (FreezyLemon) [reply osu.Game/Overlays/BeatmapSet/PreviewButton.cs]: Whoops, that was a makeshift fix for the Direct panels crashing (has already been resolved in another PR as far as I'm aware). Will remove in a second

**Comment by peppy** [comment]
Adding dependency on https://github.com/ppy/osu-framework/pull/1273.

**Comment by FreezyLemon** [comment]
As discussed, I changed the following things:
- Removed the `LinkId`
- Moved all logic à la "is this a link or not" out of `ChatLink` (which also means no parent accessors)
- Removed word wrapping by replacing spaces with `_` to avoid previous issues with hovering

Other changes:
- Changed to proper `OsuHoverContainer` usage
  - use `OsuHoverContainer.Action` to handle clicks
  - change _its_ colour instead of `ChatLink`'s **and** implemented a `OsuLinkSpriteText.TextColour` property that gets/sets `Content.Colour` -> `OsuHoverContainer.Colour` (see to-do)
  - changed `OsuHoverContainer` to fade back to the colour that is set on load instead of always white.
- moved `ChatLink.getId` into `OnLinkClicked` (as a local function) since it's only used in that method

Still to do:
- [ ] Implement word wrapping
- [ ] Move the `TextColour` property to `SpriteText`? This seems useful to avoid errors where `Content != this` and `Content` changes its colour, like here. This would usually overlap the two colours instead of properly changing.
  * Remove the cast inside `ChatLine.updateMessageContent`

File: osu.Game/Online/Chat/ChatLink.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                        switch (args[0])
+                        {
+                            case "chan":
+                                var foundChannel = chat.AvailableChannels.Find(channel => channel.Name == args[1]);
+
+                                // links should be filtered out by now if a channel doesn't exist
+                                chat.OpenChannel(foundChannel ?? throw new ArgumentException($"Unknown channel name ({args[1]})."));
+                                break;
+                            case "edit":
+                                chat.Game?.LoadEditorTimestamp();
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatLink.cs]: You should not be accessing game through chat. Please DI game if you need game.

Also you'll want to make the BDL method nullable as game won't always be available.

File: osu.Game/Online/Chat/ChatLink.cs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+                                break;
+                            case "s":
+                            case "dl":
+                                if (args.Length > 1 && int.TryParse(args[1], out int mapSetId))
+                                    beatmapSetOverlay.ShowBeatmapSet(mapSetId);
+
+                                break;
+                            case "spectate":
+                                GetUserRequest req;
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatLink.cs]: i really wouldn't bother adding the `GetUserRequest` stuff for now. it's not achieving anything.

File: osu.Game/Online/Chat/ChatLink.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+                                req.Success += user =>
+                                {
+                                    chat.Game?.LoadSpectatorScreen();
+                                };
+                                api.Queue(req);
+
+                                break;
+                            default:
+                                throw new ArgumentException($"Unknown osu:// link at {nameof(ChatLink)} ({urlMatch.Groups["content"].Value}).");
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatLink.cs]: This will cause a hard crash. Not sure an exception is what you're looking for here.

File: osu.Game/Online/Chat/ChatLink.cs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+                        break;
+                    default:
+                        base.OnLinkClicked();
+                        break;
+                }
+            }
+            else
+                base.OnLinkClicked();
+
+            int getId(string input)
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatLink.cs]: not a huge fan of this method. why not just `.TrimStart('#')`?

File: osu.Game/Online/Chat/ChatLink.cs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        }
+
+        public string TooltipText
+        {
+            get
+            {
+                if (Url == Text)
+                    return null;
+
+                var urlMatch = getUrlMatch();
```
Comment (peppy) [inline osu.Game/Online/Chat/ChatLink.cs]: really ugly that we have to do this here. going forward we will need to encapsulate the logic for each link type somehow, so it's not spread across multiple methods like this.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Online/Chat/MessageFormatter.cs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            public string Text;
+            public string OriginalText;
+
+            public MessageFormatterResult(string text)
+            {
+                OriginalText = Text = text;
+            }
+        }
+
+        public class Link : IComparable<Link>
```
Comment (peppy) [inline osu.Game/Online/Chat/MessageFormatter.cs]: Let's move this out so it's not a nested class.

File: osu.Game/OsuGame.cs [hunk 12]
```diff
@@ -115,6 +115,34 @@ private void load(FrameworkConfigManager frameworkConfig)
...
+        internal void LoadSpectatorScreen()
+        {
+            notifications.Post(new SimpleNotification
+            {
+                Text = @"Sorry, but spectating is not implemented yet!",
+                Icon = FontAwesome.fa_life_saver,
+            });
+        }
+
+        internal void LoadMultiplayerLobby(int lobbyId)
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: Should be `LoadMatch(matchId)` if anything (the lobby is a place you find many matches)

File: osu.Game/Overlays/BeatmapSetOverlay.cs [hunk 13]
```diff
@@ -139,6 +139,20 @@ protected override bool OnClick(InputState state)
             return true;
         }
 
+        public void ShowBeatmap(int beatmapId)
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapSetOverlay.cs]: this should probably be done in a separate pull request (and should have the user arrive to the correct difficulty's screen, rather than ignoring that). implementing methods like this without a TODO or an associated issue is carelessly creating bugs due to lacking implementation.

File: osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+            : base(defaultCreationParameters)
+        {
+        }
+
+        protected override SpriteText CreateSpriteText() => new T();
+
+        /// <summary>
+        /// The colour for text (links override this). Will only be used for new text elements.
+        /// </summary>
+        public ColourInfo TextColour = Color4.White;
```
Comment (peppy) [inline osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs]: why is this required?

File: osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+        public IEnumerable<SpriteText> AddLink(string text, string url, Action<SpriteText> creationParameters = null)
+        {
+            // TODO: Remove this and get word wrapping working
+            text = text.Replace(' ', '_');
+
+            return AddText(text, link =>
```
Comment (peppy) [inline osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs]: rather than calling `AddText` and making every single `SpriteText` be an `OsuLink`, why not just do the link construction here? Means you won't need all this casting, default colour logic, generic T classes etc.

File: osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs (hunk 14 above)
Comment (FreezyLemon) [reply osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs]: It's for messages with a "special" sender ('with background' in code). At the moment, the whole `contentFlow` changes colour which is not what we want since if we do that, the colours will overlap since the links will still be blue. For example: a special sender with yellow colour will have green-ish links with the old way of doing this. So the Sprites should be assigned the color directly instead of the whole container.

File: osu.Game/Overlays/BeatmapSetOverlay.cs (hunk 13 above)
Comment (FreezyLemon) [reply osu.Game/Overlays/BeatmapSetOverlay.cs]: Sorry. Got a bit too excited there and just put it in carelessly.

File: osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs (hunk 15 above)
Comment (FreezyLemon) [reply osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs]: This exists so that the link in `ProfileHeader` still works. Changing this class to creating the links itself (and to remove generics) would require a way of differentiating between general osu links and the more specific chat links. It would be easy to just put two different methods there or inherit this from another class just for chat links, but I'm really not sure what the preferred solution for this would be.

File: osu.Game/Online/Chat/ChatLink.cs (hunk 8 above)
Comment (FreezyLemon) [reply osu.Game/Online/Chat/ChatLink.cs]: Oh, right. For some reason I thought that "invalid" osu:// links would be filtered out already, my bad. Just ignoring them should be fine then.

File: osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs (hunk 15 above)
Comment (FreezyLemon) [reply osu.Game/Graphics/Containers/OsuLinkFlowContainer.cs]: I like the idea, but creating the sprite inside this means I can't use the `base.Add...` methods since all of them create the SpriteTexts based on the input text. If I want to do this differently I'd have to change `TextFlowContainer.Add` to not throw an exception anymore, and that doesn't seem intended. But if I changed it so we would _not_ let this class inherit from `TextFlowContainer`, a lot of the code would probably have to be rewritten

**Comment by FreezyLemon** [comment]
How does the following idea sound:
Instead of letting the `ChatLink` itself decide on what to do (when it's clicked), how about the `MessageFormatter` passing the `Link` an "Action" or "Type" or something (enum) with an argument (string) based on the URL and used regex? That way we would avoid having to run the same kind of pattern matching multiple times.

The sprite then only has to decide on an action based on an enum and then use the argument in said action.
For example: Action `LinkAction.OpenBeatmapSet` with the argument "93523". The `ChatLink` only has to parse the string to an int and can then pass the mapset ID to the beatmap overlay which can then open.

**Comment by peppy** [comment]
Something has regressed in layout: https://streamable.com/ouzbx

Notice the fill flow flashing weirdly.

**Comment by FreezyLemon** [comment]
Okay, so:
The `ChatLink.Content` (an `OsuHoverContainer` instance) doesn't update its `DrawSize` correctly.
It seems that the actual `Size` is correct, but the DrawSize isn't invalidated until `loadComplete` automatically invalidates it again. It might also be something with wrong `RequiredParentSizeToFit` values of the children, but I'm not sure about that.

**Comment by peppy** [comment]
Links don't seem to work any more :(

**Comment by FreezyLemon** [comment]
What link didn't work for you? Seems like everything still works to me..
(Note that in tests, only external links have a visible effect since the DI'd game instance doesn't exist)

**Comment by FreezyLemon** [comment]
- [ ] ppy/osu-framework#1324

That fixes the tooltips not working. But I still can't reproduce what you're saying, links are working for me.

**Comment by FreezyLemon** [comment]
Oh, I see what you mean now. Investigating..

**Comment by peppy** [comment]
i still need to do a pass on code quality, but also found:

- osump:// links don't seem to work (and aren't tested).
- osu:// links don't seem to work (and aren't tested).
- typing an editor link gobbles the portion after the

**Comment by peppy** [comment]
All aforementioned issues are now fixed. I'm going to do a second pass over the code as I'm still not confident about some of what is going on here.

Safe to say it's in a working and tested state now, though.

**Comment by peppy** [comment]
I think I'm happy with this one. @smoogipoo would you like the opportunity to review this before merging? I rewrote a good portion so if you feel the need to review then let me know and I'll wait for that to happen.

**Comment by smoogipoo** [comment]
Yes, I would like to review this.

**Review by peppy — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]
